api/v1/storageprofile_types.go

```diff
@@ -24,25 +24,25 @@ import (
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
 // StorageProfileSpec defines the desired state of StorageProfile
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.deviceClass) || has(self.deviceClass)", message="deviceClass is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.sharedFilesystemConfiguration) || has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.deviceClass) && has(self.deviceClass)", message="deviceClass is required"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.sharedFilesystemConfiguration) && has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.blockPoolConfiguration) && has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required"
 type StorageProfileSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="DeviceClass is immutable"
 	// +kubebuilder:validation:MaxLength=512
 	// DeviceClass is the deviceclass name.
 	DeviceClass string `json:"deviceClass"`
 
-	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Required
```

Comment:
AFAIK this is optional not a mandatory to have it

Reply:
The idea is to set all fields to required. The issue before was that the fields could've been initially unset and we aim to have them set at all times.

Reply:
@Madhu-1 The requirement was for immutable CRs. The current implementation allows you to set values to unset fields after creation. I believe @raaizik fix will achieve that

Reply:
@Madhu-1 I think you meant that it's uncertain whether the other fields, apart from `deviceClass`, should be mandatory _outside_ of the fact that currently the values can be initially unset before set, @nb-ohad. I think we made an assumption that they should be mandatory, not sure why though.

Reply:
@raaizik I think you are right! I just observed it myself. 
We don't want to require the customer to fill in all of the fields. We just want that the Spec will stay immutable after CR creation. 

I think what we need is a validation rule on the Spec field itself that forces it to be identical to the old spec field

Reply:
@Madhu-1 The CEL rules prevent updates and removals. The difference is [this](https://github.com/red-hat-storage/ocs-operator/pull/2231#discussion_r1371764253)

Reply:
To prove my case, I was able to create an SP without `blockPoolConfiguration` and then edit it to include it! In order to enforce the immutability restriction mentioned, I'd need to do it on a per sub-field basis per each complex type field.

Reply:
- you mean setting validation on `type SharedFilesystemConfigurationSpec struct` and `type BlockPoolConfigurationSpec struct`, that might interfere with StorageCluster Spec, isn't it?
- btw, I don't get why `omitempty` is in the picture, I believe it doesn't help with validations but only for (de/)serialize (or) you want to try this combo with `deviceClass` as optional?

Reply:
1. No.
2. It's pretty self explanatory. It literally allows optional fields that are left blank to be omitted (i.e., all the non `deviceClass` fields).